Update permissions documentation (#1141)
bolt/docs: docs/configuration/permissions.md

Bolt uses a [Role-Based Permission](https://en.wikipedia.org/wiki/Role-based_access_control)
system. This means that:

> * Every *user* can have zero or more *roles*

Reviewer: I think "can have" can be replaced with "has". It's not optional to have 'zero or more roles'
```suggestion
* Every *user* has zero or more *roles*
```

* Every *role* can grant zero or more *permissions*
* The same *permission* can be granted through several different *roles*

> * Several users can have the same *role*

Reviewer: ```suggestion
* Several users can have the same *role*, and one role can be granted to multiple *users*
```

 1. Permissions are quite central to Bolt's inner workings, and by
    misconfiguring them, you can lock yourself out - for example, removing the
    `IS_AUTHENTICATED_REMEMBERED` role from the `dashboard` permission will make

>     the bolt dashboard completely inaccessible for everyone.

Reviewer: ```suggestion
    the Bolt dashboard completely inaccessible for everyone.
```

 2. If somebody has the permission to `user:edit`, they can also grant
    permissions to themselves or others. This means they can give themselves
    roles with more permissions, or take away those roles from others.
    This is similar for a lot of permissions that are by default assigned
    to ROLE_ADMIN only.
    In short: **Do _not_ give someone the `ROLE_ADMIN` (or `ROLE_DEVELOPER`) role, 
    or change the setup in a way that they have permission to do things that were

>     previously only accessible to ROLE_ADMIN unless you trust them fully!!**

Reviewer: ```suggestion
    previously only accessible to `ROLE_ADMIN` unless you trust them fully!**
```

Configuring Roles
-----------------
Bolt uses [Symfony Security](https://symfony.com/doc/current/security.html)
for its Roles and Permissions setup. Because of this part of the security configuration
is in `config/packages/security.yaml`.

The default setup will contain the following assignable roles: `ROLE_DEVELOPER`, `ROLE_ADMIN`, 
`ROLE_CHIEF_EDITOR`, `ROLE_EDITOR`, `ROLE_USER`.

These roles are put in a hierarchy specified in `config/packages/security.yaml`, this 
means that in the order specified above every role also has the permissions given to
the roles that are following it. (So `ROLE_ADMIN` can do everything `ROLE_CHIEF_EDITOR` can do etc.)

Below is an explanation of permissions for these roles in general terms, as
configured by default. Always check `config/bolt/permissions.yaml` if you want
to be certain.

`ROLE_USER` is the role a user gets when there is no role set. It doesn't

> give access to anything in the bolt admin.

Reviewer: ```suggestion
give access to anything in the Bolt administration panel.
```

as a second parameter, e.g. `is_granted('change-status', record)` 

A few examples:

```php
# check if the currently authenticated user has permission to view the dashboard
/** @var AuthorizationCheckerInterface $authorizationChecker */
$authorizationChecker->isGranted('dashboard');
```

```twig
# check if the currently authenticated user can change the status of this record:
is_granted('change-status', record)
```

Debugging Permissions
---------------------

Use the symfony toolbar to debug permissions in combination with the 'switch user'
functionality available to users with the `ROLE_DEVELOPER` role.

Extending the Permission System
---------------------

Bolt uses [Symfony Security](https://symfony.com/doc/current/security.html), this
makes it relatively easy to obtain documentation and tips on how to achieve what you want.
Start by reading the linked documentation when you want to make a change.
The Bolt classes that implement the Bolt specific functionality are 
in the `Bolt\Security` namespace. `ContentVoter` and `GlobalVoter` are doing most
of the decision making. (Strictly speaking they are only 'voting' and the decision is
made elsewhere in the security system)

> If you search the bolt core code for `isGranted` and the templates for `is_granted` you will 

Reviewer: ```suggestion
If you search the Bolt core code for `isGranted` and the templates for `is_granted` you will 
```
